linkerd/stack/src/gate.rs

```diff
@@ -1,81 +1,97 @@
 use crate::Service;
 use futures::{ready, FutureExt};
 use std::{
-    sync::{atomic::AtomicBool, Arc},
+    sync::Arc,
     task::{Context, Poll},
 };
-use tokio::sync::Notify;
+use tokio::sync::{watch, OwnedSemaphorePermit, Semaphore, TryAcquireError};
 use tokio_util::sync::ReusableBoxFuture;
-use tracing::{debug, trace};
+use tracing::debug;
 
 /// A middleware that alters its readiness state according to a gate channel.
 pub struct Gate<S> {
     inner: S,
     rx: Rx,
-    is_waiting: bool,
-    waiting: ReusableBoxFuture<'static, ()>,
+    acquire: ReusableBoxFuture<'static, Option<OwnedSemaphorePermit>>,
+    acquiring: bool,
+    permit: Poll<Permit>,
 }
 
 /// Observes gate state changes.
 #[derive(Clone, Debug)]
-pub struct Rx(Arc<Shared>, Arc<()>);
+pub struct Rx(watch::Receiver<State>);
 
 /// Changes the gate state.
 #[derive(Clone, Debug)]
-pub struct Tx(Arc<Shared>);
+pub struct Tx(Arc<watch::Sender<State>>);
 
-#[derive(Debug)]
-struct Shared {
-    open: AtomicBool,
-    notify: Notify,
-    closed: Notify,
+#[derive(Clone, Debug)]
+pub enum State {
+    /// The gate is open and unlimited.
+    Open,
+
+    /// The gate is limited by the provided semaphore. Permits are forgotten as
+    /// requests are processed so that this semaphore can specify a limit on the
+    /// number of requests to be admitted by the [`Gate`].
+    Limited(Arc<Semaphore>),
+
+    /// The gate is shut and no requests are to be admitted.
+    Shut,
 }
 
 /// Creates a new gate channel.
 pub fn channel() -> (Tx, Rx) {
-    let shared = Arc::new(Shared {
-        open: AtomicBool::new(true),
-        notify: Notify::new(),
-        closed: Notify::new(),
-    });
-    (Tx(shared.clone()), Rx(shared, Arc::new(())))
+    let (tx, rx) = watch::channel(State::Open);
+    (Tx(Arc::new(tx)), Rx(rx))
 }
 
+struct Permit(Option<OwnedSemaphorePermit>);
+
 // === impl Rx ===
 
 impl Rx {
     /// Indicates whether the gate is open.
+    pub fn state(&self) -> State {
+        self.0.borrow().clone()
+    }
+
```

Comment:
tiny comment nit, i think the comment "Indicates whether the gate is open" is meant to be on `is_open`:
```suggestion
    pub fn state(&self) -> State {
        self.0.borrow().clone()
    }
    
    /// Indicates whether the gate is open.
```